amr-wind/fvm/nonLinearSum.H

```diff
@@ -0,0 +1,161 @@
+                                     (ux * wx + uy * wy + uz * wz) +
+                                     3 * (ux * uz + vx * vz + wx * wz) +
+                                     (ux * wx + vx * wy + wx * wz);
+                // N21
+                strphi(i, j, k, 3) = (vx * ux + vy * vx + vz * wx) -
+                                     (vx * ux + vy * uy + vz * uz) +
+                                     3 * (uy * ux + vy * vx + wy * wx) +
+                                     (uy * ux + vy * uy + wy * uz);
+                // N22
+                strphi(i, j, k, 4) = (vx * uy + vy * vy + vz * wy) -
+                                     (vx * vx + vy * vy + vz * vz) +
+                                     3 * (uy * vx + vy * vy + wy * wy) +
+                                     (uy * vx + vy * vy + wy * vz);
+                // N23
+                strphi(i, j, k, 5) = (vx * uz + vy * vz + vz * wz) -
+                                     (vx * wx + vy * wy + vz * wz) +
+                                     3 * (uy * wx + vy * vz + wy * wz) +
+                                     (uy * wx + vy * wy + wy * wz);
+                // N31
+                strphi(i, j, k, 6) = (wx * ux + wy * vx + wz * wx) -
+                                     (wx * ux + wy * uy + wz * uz) +
+                                     3 * (uz * ux + vz * vx + wz * wx) +
+                                     (uz * ux + vz * uy + wz * uz);
+                // N32
+                strphi(i, j, k, 7) = (wx * uy + wy * vy + wz * wy) -
+                                     (wx * vx + wy * vy + wz * vz) +
+                                     3 * (uz * uy + vz * vy + wz * wy) +
+                                     (uz * vx + vz * vy + wz * vz);
+                // N33
+                strphi(i, j, k, 8) = (wx * uz + wy * vz + wz * wz) -
+                                     (wx * wx + wy * wy + wz * wz) +
+                                     3 * (uz * uz + vz * vz + wz * wz) +
+                                     (uz * wx + vz * wy + wz * wz);
+            });
+    }
+
+    FTypeOut& m_strphi;
+    const FTypeIn& m_phi;
+};
+
+/** Compute the non-linear term for the Kosovic Sub-grid scale model
+ *  \ingroup fvm
+ *
+ *  \param strphi [out] Field where the non-linear terms is populated
+ *  \param phi [in] The stress tensor  field
+ */
+template <typename FTypeIn, typename FTypeOut>
+inline void nonlinearsum(FTypeOut& strphi, const FTypeIn& phi)
+{
+    BL_PROFILE("amr-wind::fvm::nonlinearsum");
+    NonLinearSum<FTypeIn, FTypeOut> str(strphi, phi);
+    impl::apply(str, phi);
+}
+
+/** Compute the non-linear term as a ScratchField
+ *  \ingroup fvm
+ *
+ *  \param phi [in] The stress tensor  field
+ */
+template <typename FType>
+inline std::unique_ptr<ScratchField> nonlinearsum(const FType& phi)
+{
+    const std::string gname = phi.name() + "_nonlinearsum";
+    auto strphi = phi.repo().create_scratch_field(gname, 1);
```

Comment:
I think I found the error! This scratch field is initialized with only 1 component, but you appear to need 9 components for what goes on in the operator.